lib/Dialect/FIRRTL/Transforms/SpecializeLayers.cpp

```diff
@@ -0,0 +1,725 @@
+      instanceChoice->erase();
+      instanceChoice = newInstanceChoice;
+    }
+
+    // Specialize the required enable layers.  Due to the layer verifiers, there
+    // should not be any disabled layer in this instanceChoice and this should
+    // be infallible.
+    auto newLayers = specializeEnableLayers(instanceChoice.getLayersAttr());
+    instanceChoice.setLayersAttr(newLayers.getValue());
+
+    insertionPoint.moveOpBefore(instanceChoice);
+  }
+
+  void specializeOp(WireOp wire, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    if (specializeValue(wire.getResult())) {
+      insertionPoint.moveOpBefore(wire);
+    } else {
+      if (auto innerSym = wire.getInnerSymAttr())
+        recordRemovedInnerSym(removedSyms,
+                              wire->getParentOfType<FModuleOp>().getNameAttr(),
+                              innerSym);
+      wire.erase();
+    }
+  }
+
+  void specializeOp(RefDefineOp refDefine, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    // If this is connected disabled probes, erase the refdefine op.
+    if (specializeLayerRef(refDefine.getDest().getType().getLayer()))
+      insertionPoint.moveOpBefore(refDefine);
+    else
+      refDefine->erase();
+  }
+
+  void specializeOp(RefSubOp refSub, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    if (specializeValue(refSub->getResult(0)))
+      insertionPoint.moveOpBefore(refSub);
+    else
+      refSub.erase();
+  }
+
+  void specializeOp(RefCastOp refCast, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    if (specializeValue(refCast->getResult(0)))
+      insertionPoint.moveOpBefore(refCast);
+    else
+      refCast.erase();
+  }
+
+  void specializeOp(Operation *op, InsertionPoint &insertionPoint,
+                    DenseSet<Attribute> &removedSyms) {
+    TypeSwitch<Operation *>(op)
+        .Case<LayerBlockOp, WhenOp, MatchOp, InstanceOp, InstanceChoiceOp,
+              WireOp, RefDefineOp, RefSubOp, RefCastOp>(
+            [&](auto op) { specializeOp(op, insertionPoint, removedSyms); })
+        .Default([&](Operation *op) {
+          // By default all operations should be inlined from an enabled
+          // layer.
+          insertionPoint.moveOpBefore(op);
+        });
+  }
+
+  /// Specialize a block of operations, removing any probes which are
+  /// disabled, and moving all operations to the insertion point.
+  void specializeBlock(Block *block, InsertionPoint &insertionPoint,
+                       DenseSet<Attribute> &removedSyms) {
+    // Since this can erase operations that deal with disabled probes, we walk
+    // thee block in reverse to make sure that we erase uses before defs.
+    for (auto &op : llvm::make_early_inc_range(llvm::reverse(*block)))
+      specializeOp(&op, insertionPoint, removedSyms);
```

Comment:
This made me very confused for a while because I thought this was doing function overloading _at runtime_... I missed the `specializedOp` with the `TypeSwitch` above.

Personally, this would be more readable with that function inlined here as the indirection to only a `TypeSwitch` seems a bit too finely grained.